otaserver/coap.py

```diff
@@ -64,24 +97,48 @@ def __init__(self, fw_path, port=COAP_PORT):
         self.port = port
         self.fw_path = fw_path
         root_coap = resource.Site()
-        root_coap.add_resource(('version', ), FirmwareVersionResource(self))
-        root_coap.add_resource(('firmware', ), FirmwareBinaryResource(self))
+        self.appids = {}
+        firmwares = os.listdir(self.fw_path)
+        for fw in firmwares:
+            firmware = FirmwareUpdate(fw)
+            appid = fw.split("-")[-2]
+            if appid in self.appids:
+                self.appids[appid].append(fw)
+                continue
+            self.appids[appid] = [fw]
+            print (appid)
+            root_coap.add_resource((appid, 'version', ), FirmwareVersionResource(self, appid))
+            root_coap.add_resource((appid, 'slot1', 'name', ), FirmwareNameResource(self, appid, 1))
+            root_coap.add_resource((appid, 'slot2', 'name', ), FirmwareNameResource(self, appid, 2))
+        
         asyncio.async(Context.create_server_context(root_coap,
                                                     bind=('::', self.port)))
 
-    def get_latest_version(self):
+    def get_latest_version(self, appid):
         """Get the latest firmware version."""
         firmwares = os.listdir(self.fw_path)
         if firmwares == []:
             return ''
         else:
-            return '.'.join(max(os.path.splitext(
-                fw)[0].split("-")[-1].split(".") for fw in firmwares))
+            fw_appids = list(filter(lambda x: x.split("-")[-2] == appid, firmwares))
+            return max([fw.split("-")[-1] for fw in fw_appids])
+
+    def get_latest_fw_name(self, appid, slot):
+        """Get the latest firmware name."""
+        firmwares = os.listdir(self.fw_path)
+        if firmwares == []:
+            return ''
+        else:
+            version = self.get_latest_version(appid)
+            fw_appids = list(filter(lambda x: x.split("-")[-2] == appid and x.split("-")[-1] == version, firmwares))
+            print (fw_appids)
+            fw_name = list(filter(lambda x: x.split("-")[-3] == 'slot' + str(slot), fw_appids))
```

Comment:
I'm just realizing that the logic of this line is duplicated a lot of time. It can surely be factorized within a dedicated helper function (with appropriate parameters)